stm32n6: enable SMH on PSRAM & add video sample conf files

avolmat-st opened the pull request:
This PR add the support for the SMH (Shared Multi Heap) on the XSPI PSRAM in order to store LTDC & video buffers.
On top of that stm32n6 board configuration files are added in order to use sample/video/capture and sample/video/capture_to_lvgl on the stm32n657_dk board with st_b_cams_imx_mb1854 shield

Tested with:
west build -p -b stm32n6570_dk --shield st_b_cams_imx_mb1854 samples/drivers/video/capture
west build -p -b stm32n6570_dk --shield st_b_cams_imx_mb1854 samples/drivers/video/capture_to_lvgl

In order to have the sample app working, the PR #79996 is also needed in order to have the capture sample app avoid requesting BGR display format instead of RGB display format.

Commit:
memc: stm32_xspi_psram: init shared_multi_heap area

Initialize the whole psram as a shared_multi_heap_area
if SHARED_MULTI_HEAP is enabled.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
samples: video: capture: add stm32n6570_dk config files

Add capture sample config files for the stm32n6570_dk board.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
samples: video: capture_to_lvgl: use video_buffer_aligned_alloc

Use video_buffer_aligned_alloc instead of video_buffer_alloc for buffer
allocation in order to be able to enforce required buffer alignment
configured via CONFIG_VIDEO_BUFFER_POOL_ALIGN.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
samples: video: capture_to_lvgl: add stm32n6570_dk conf files

Add configuration files for the stm32n6570_dk board.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
display: stm32: ltdc: add SMH buffer alignment config option

Add the CONFIG_STM32_LTDC_FB_SMH_ALIGN option in order to
indicate the alignment required for the buffer allocated via
SMH.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

erwango commented on samples/drivers/video/capture/boards/stm32n6570_dk.conf:
```diff
@@ -0,0 +1,15 @@
+CONFIG_VIDEO_BUFFER_POOL_SZ_MAX=800000
+CONFIG_VIDEO_BUFFER_POOL_NUM_MAX=2
+CONFIG_VIDEO_BUFFER_POOL_ALIGN=16
+CONFIG_MULTI_HEAP=y
+CONFIG_SHARED_MULTI_HEAP=y
+CONFIG_VIDEO_BUFFER_USE_SHARED_MULTI_HEAP=y
+CONFIG_VIDEO_BUFFER_SMH_ATTRIBUTE=2
+CONFIG_STM32_LTDC_FB_USE_SHARED_MULTI_HEAP=y
+CONFIG_STM32_LTDC_FB_SMH_ATTRIBUTE=2
+CONFIG_STM32_LTDC_FB_SMH_ALIGN=256
```
Can we put most of those in a more systematic way so either SoC or board default config depending on the combination of IP in use ?

avolmat-st replied:
I think it depends on what the application is at the end. By default the ltdc has the ext-sdram property so there is not real need (by default) to have the SMH for the LTDC part. However as soon as we had the camera (via the st_b_cams_imx shield) there is a very good chance that the video frame won't fit into the SRAM so PSRAM (via SMH) is needed. In which case, LTDC (if needed) and video buffer should be put via SMH.
Maybe I could add this into the shield configuration when running on the stm32n6 then. What do you think ?

Commit:
soc: st: stm32: set default SMH attribute for LTDC/video buffers

The SMH attribute when using the XSPI PSRAM is set to EXTERNAL (2)
within the driver hence set default for both LTDC and video
buffer SMH attribute to 2 if all conditions are validated.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
boards: stm32n6570_dk: enable SMH by default for LTDC/VIDEO

By default, enable the SharedMultiHeap allocated buffer on
stm32n6570_dk and enable its usage for LTDC or video buffers.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
soc: st: stm32: stm32n6: set 256 SMH buffer alignment for LTDC

Set the LTDC buffer alignment to 256 in order to avoid an
issue when accessing to PSRAM via XSPI.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
soc: st: stm32: set default video buffer align to 16 for DCMIPP

Set the default video buffer alignment constraint to 16 when DCMIPP
is being used.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

josuah commented on drivers/memc/memc_stm32_xspi_psram.c:
```diff
@@ -14,11 +14,20 @@
 
+#ifdef CONFIG_SHARED_MULTI_HEAP
+struct shared_multi_heap_region smh_psram = {
+	.addr = DT_REG_ADDR(DT_NODELABEL(psram)),
+	.size = DT_REG_SIZE(DT_NODELABEL(psram)),
+	.attr = SMH_REG_ATTR_EXTERNAL,
```
A comparison with Espressif for pure curiosity:

https://github.com/zephyrproject-rtos/zephyr/blob/8ab712a2353862277fb23191a5f77e35057c3e63/soc/espressif/esp32s2/memory.h#L61-L62

https://github.com/zephyrproject-rtos/zephyr/blob/8ab712a2353862277fb23191a5f77e35057c3e63/soc/espressif/esp32/default.ld#L32-L35

https://github.com/zephyrproject-rtos/zephyr/blob/8ab712a2353862277fb23191a5f77e35057c3e63/soc/espressif/esp32/default.ld#L850-L853

https://github.com/zephyrproject-rtos/zephyr/blob/8ab712a2353862277fb23191a5f77e35057c3e63/soc/espressif/common/esp_psram.c#L17-L21

josuah commented on soc/st/stm32/Kconfig.defconfig:
```diff
@@ -85,4 +85,12 @@ config FLASH_BASE_ADDRESS
 	default $(dt_chosen_reg_addr_hex,$(DT_CHOSEN_Z_FLASH))
 
+config STM32_LTDC_FB_SMH_ATTRIBUTE
+	default 2 if MEMC_STM32_XSPI_PSRAM && SHARED_MULTI_HEAP && \
+		     STM32_LTDC_FB_USE_SHARED_MULTI_HEAP
```
[non-blocking]
It could be great to recall what `2` means, eventually defining this in `lib/heap/Kconfig`:

```Kconfig
SMH_REG_ATTR_CACHEABLE := 0
SMH_REG_ATTR_NON_CACHEABLE := 1
SMH_REG_ATTR_EXTERNAL := 2
```

But this is beyond the scope of this PR.

avolmat-st replied:
Added a commend on top of those lines to indicate what is this 2.

josuah commented on boards/st/stm32n6570_dk/Kconfig.defconfig:
```diff
@@ -12,19 +12,35 @@ config NET_L2_ETHERNET
 endif # NETWORKING
 
-if DISPLAY
+if DISPLAY || VIDEO
+# MEMC needs to be enabled in order to store
+# display frame buffer to external PSRAM
+config MEMC
+	default y
+
+# Rely on Multi-Heap / Shared-Multi-Heap for PSRAM access
+config MULTI_HEAP
+	default y
+config SHARED_MULTI_HEAP
+	default y
+endif # DISPLAY || VIDEO
```
[non-blocking]
Is this to be defined at board-level or SoC-level?

josuah reviewed (commented):
Glad to see SMH used more! Some questions and comments below but LGTM AFAICT!

erwango commented on soc/st/stm32/Kconfig.defconfig:
```diff
@@ -85,4 +85,12 @@ config FLASH_BASE_ADDRESS
 		if $(DT_FLASH_PARENT_IS_XSPI)
 	default $(dt_chosen_reg_addr_hex,$(DT_CHOSEN_Z_FLASH))
 
+config STM32_LTDC_FB_SMH_ATTRIBUTE
```
Please add a short comment so reader can have an idea of why this is needed (and meaningful to put it here)

erwango commented on soc/st/stm32/Kconfig.defconfig:
```diff
@@ -85,4 +85,12 @@ config FLASH_BASE_ADDRESS
 		if $(DT_FLASH_PARENT_IS_XSPI)
 	default $(dt_chosen_reg_addr_hex,$(DT_CHOSEN_Z_FLASH))
 
+config STM32_LTDC_FB_SMH_ATTRIBUTE
```
Nitpick: To avoid repetition of the condition, put both definitions under:
```
if MEMC_STM32_XSPI_PSRAM && SHARED_MULTI_HEAP && \
		     STM32_LTDC_FB_USE_SHARED_MULTI_HEAP
endif
```

avolmat-st replied:
Put both into a if section
+if MEMC_STM32_XSPI_PSRAM && SHARED_MULTI_HEAP

and then define the default depending if SMH is enabled or not for LTDC or VIDEO_BUFFER

erwango reviewed (changes requested):
2 comments, otherwise LGTM for STM32 side of things.
